Feat(eos_designs): Only configure vlans in use by connected endpoints or downstream L2 switches

ClausHolbechArista opened the pull request:
## Change Summary

<!-- Enter short PR description -->
Only configure vlans in use by connected endpoints or downstream L2 switches

## Related Issue(s)

Depends on PR #1826, so this must be rebased after 1826 has merged.

@perimore does this address #1575 ?

Fixes #1822 

## Component(s) name

`arista.avd.eos_designs`

## Proposed changes
<!--- Describe your changes in detail -->
<!--- Describe data model implemented for new features -->

### Fabric topology vars

```
< node_type_key >:
  defaults:
    filter:
      # Only configure VLANs, SVIs, VRFs in use by connected endpoints or downstream L2 switches.
      # Note! This feature only considers configuration managed by eos_designs.
      # This excludes structured_config, custom_structured_configuration_, raw_eos_cli, eos_cli, custom templates, configlets etc.
      only_vlans_in_use: < true | false | default -> false >
```

### Fabric vars

```
# Only Configure VLAN Trunk Groups used by local endpoints | Optional
# A vlan can have many trunk_groups assigned. To avoid unneeded configuration changes on all leaf
# switches when a new trunk group is added, this feature will only configure the vlan trunk groups
# matched with local connected_endpoints.
# See "Details on only_local_vlan_trunk_groups" below.
# Requires "enable_trunk_groups: true"
only_local_vlan_trunk_groups: < true | false | default -> false >
```

## How to test
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->

Added molecule coverage:
- Tested both with and without trunk groups and with and without mlag.
- Tested both features together and seperately.

## Checklist

### User Checklist

<!-- Add your own checklist using MD syntax and by replacing N/A -->
- N/A

### Repository Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

perimore commented:
This partially addresses #1575 with respect to L2 filtering but #1575 also calls for the pruning of VLANs and VRFs from an L3 device if there are no endpoints requiring the services.

ClausHolbechArista committed:
Feat(eos_designs): Only configure vlans in use by connected endpoints

ClausHolbechArista committed:
Add test for filter.only_vlans_in_use without trunk groups

ClausHolbechArista committed:
Improve docs

ClausHolbechArista commented:
> This partially addresses #1575 with respect to L2 filtering but #1575 also calls for the pruning of VLANs and VRFs from an L3 device if there are no endpoints requiring the services.

@perimore re. #1575 I believe this PR addresses all of those points. It works on all devices that support "connected_endpoints" so `l3leaf` and `l2leaf` by default. It will both filter on the device's own `connected_endpoints` as well as any downstream devices.
See [this example](https://github.com/aristanetworks/ansible-avd/pull/1821/files#diff-b5119ecbc315e6623102a1674a7308141cd3dbd87d88d2fcc58965406b485327), where only two server vlans are configured, and all other vlans+vrfs+vnis are pruned away.

I would appreciate if you could find time to test this out.

jonxstill commented on ansible_collections/arista/avd/plugins/module_utils/eos_designs_facts.py:
```diff
@@ -651,13 +657,114 @@ def virtual_router_mac_address(self):
+                        continue
+
+                    if "vlans" in adapter_settings and adapter_settings["vlans"] not in ["all", "", None]:
+                        vlans.extend(map(int, self._range_expand(adapter_settings["vlans"])))
```
One of my topologies failed here due to `vlans` in my combined adapter settings being an int, not a string or list, which caused range_expand to choke on it.
```
ansible.errors.AnsibleFilterError: value must be of type list or str, got <class 'int'>
```
Easy fix would be:

```suggestion
                        vlans.extend(map(int, self._range_expand(str(adapter_settings["vlans"]))))
```
But I'm not sure if want to support lists of ranges in something as simple as a VLAN range.

jonxstill reviewed (dismissed):
Other than the type issue, this looks good to me.

lermilov commented:
Hi Claus! Cool PR. I have tested it with my own repos.
Few observations:
- SVIs are also removed from L3LEAF if they are not configured in the connected endpoints. I assume this is intentional, but better to document it.
- Interfaces/VLANs configured with structured_configuration at hostvar level are ignored (Carl calls it "Port-centric" approach =)). I think it is intentional too, but let's also document it.

carlbuchmann commented:
@lermilov - re: "Interfaces/VLANs configured with structured_configuration at hostvar level are ignored (Carl calls it "Port-centric" approach =)). I think it is intentional too, but let's also document it."

I refer to port-centric when configuring network ports i.e `custom_structured_configuration_ethernet_interfaces`.

And yes, let's document this behavior and mention that custom_structured_configuration for interfaces and VLANs will be ignored.

ClausHolbechArista committed:
Update ansible_collections/arista/avd/plugins/module_utils/eos_designs_facts.py

Co-authored-by: Jon Still <jonstill@arista.com>

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/doc/fabric-topology.md:
```diff
@@ -369,6 +369,10 @@ default_interfaces:
       # Useful for "border" leaf.
       always_include_vrfs_in_tenants: [ < tenant_1 >, < tenant_2 >, "all" ]
 
+      # Only configure VLANs in use by connected endpoints
+      # or downstream L2 switches.
```
```suggestion
      # Only configure VLANs, SVIs, VRFs in use by connected endpoints
      # or downstream L2 switches.
```

ClausHolbechArista committed:
Update ansible_collections/arista/avd/roles/eos_designs/doc/fabric-topology.md

ClausHolbechArista commented:
> @lermilov - re: "Interfaces/VLANs configured with structured_configuration at hostvar level are ignored (Carl calls it "Port-centric" approach =)). I think it is intentional too, but let's also document it."
> 
> I refer to port-centric when configuring network ports i.e `custom_structured_configuration_ethernet_interfaces`.
> 
> And yes, let's document this behavior and mention that custom_structured_configuration for interfaces and VLANs will be ignored.

Anything implemented with custom_structured_configuration is not available to eos_designs. That goes for all features and is described under custom_structured_configuration.
I will add a general note about this only considering configuration managed by eos_designs. So excluding structured_config, custom_structured_config, raw_eos_cli, eos_cli and other custom configurations.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/doc/fabric-topology.md:
```diff
@@ -369,6 +369,10 @@ default_interfaces:
       # Useful for "border" leaf.
       always_include_vrfs_in_tenants: [ < tenant_1 >, < tenant_2 >, "all" ]
 
+      # Only configure VLANs, SVIs, VRFs in use by connected endpoints
+      # or downstream L2 switches.
```
```suggestion
      # Only configure VLANs, SVIs, VRFs in use by connected endpoints or downstream L2 switches.
      # Note! This feature only considers configuration managed by eos_designs.
      # This excludes structured_config, custom_structured_config, raw_eos_cli, eos_cli, custom tempaltes, configlets etc.
```

ClausHolbechArista committed:
Update ansible_collections/arista/avd/roles/eos_designs/doc/fabric-topology.md

ClausHolbechArista committed:
Fix typo in doc

nielsjlarsen reviewed (approved):
LGTM!

carlbuchmann reviewed (approved):
LGTM - tested with customer scenario.
